### PR TITLE
[broker][admin]Add api for update topic properties

### PR DESCRIPTION
**Pull request by AnonHxy**
### Motivation

* Maybe this modification should be part of  https://github.com/apache/pulsar/issues/12629 .  As described in the PIP document "These metadata could be set during topic creation and also updated.",  we have implemented the `creation` method, but have not implement  the `updated` method. 

* This PR will implement the topic meta update

### Modifications

* Add a admin api `updateProperties`

### Verifying this change

- [x] Make sure that the change passes the CI checks.
- Add UT `org.apache.pulsar.broker.admin.AdminApi2Test#testUpdatePartitionedTopicProperties`
- Add UT `org.apache.pulsar.broker.admin.AdminApi2Test#testUpdateNonPartitionedTopicProperties`

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java [hunk 1]
```diff
@@ -966,6 +966,42 @@ public void getProperties(
                 });
     }
 
+    @PUT
+    @Path("/{tenant}/{namespace}/{topic}/properties")
+    @ApiOperation(value = "Replace all the properties on the given subscription")
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java:971]: `Replace all the properties`  -> `Update the properties`

**Review by Technoboy- — APPROVED** [review]
LGTM

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java (hunk 1 above)
Comment (AnonHxy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java]: Nice catch!  Updated

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 2]
```diff
@@ -18,6 +18,7 @@
  */
 package org.apache.pulsar.broker.admin.impl;
 
+import static org.apache.commons.collections4.MapUtils.putAll;
```
Comment (Jason918) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:21]: Better avoid import static method.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 3]
```diff
@@ -601,6 +602,60 @@ private CompletableFuture<Map<String, String>> getPropertiesAsync() {
         });
     }
 
+    protected CompletableFuture<Void> internalUpdatePropertiesAsync(boolean authoritative,
+                                                                    Map<String, String> properties) {
+        if (properties == null || properties.isEmpty()) {
+            log.warn("[{}] [{}] properties is empty, ignore update", clientAppId(), topicName);
+            return CompletableFuture.completedFuture(null);
```
Comment (Jason918) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:609]: If so, is there a way to delete the properties?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java [hunk 4]
```diff
@@ -966,6 +966,42 @@ public void getProperties(
...
+        @ApiParam(value = "Specify the tenant", required = true)
+        @PathParam("tenant") String tenant,
+        @ApiParam(value = "Specify the namespace", required = true)
+        @PathParam("namespace") String namespace,
+        @ApiParam(value = "Specify topic name", required = true)
+        @PathParam("topic") @Encoded String encodedTopic,
+        @ApiParam(value = "Whether leader broker redirected this call to this broker. For internal use.")
+        @QueryParam("authoritative") @DefaultValue("false") boolean authoritative,
+        @ApiParam(value = "Key value pair properties for the topic metadata") Map<String, String> properties){
+        validatePersistentTopicName(tenant, namespace, encodedTopic);
```
Comment (Jason918) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java:993]: I wonder what happens if this validation fails.

File: pulsar-client-admin-api/src/main/java/org/apache/pulsar/client/admin/Topics.java [hunk 5]
```diff
@@ -736,6 +736,24 @@ void updatePartitionedTopic(String topic, int numPartitions, boolean updateLocal
      */
     CompletableFuture<Map<String, String>> getPropertiesAsync(String topic);
 
+    /**
+     * Update Topic Properties on a topic.
+     * The new properties will override the existing values, properties that are not passed will be keep.
+     * @param topic
```
Comment (Jason918) [inline pulsar-client-admin-api/src/main/java/org/apache/pulsar/client/admin/Topics.java:742]: > "properties that are not passed will be keep." 

Not quite clear about what this means.

Comment (AnonHxy) [reply pulsar-client-admin-api/src/main/java/org/apache/pulsar/client/admin/Topics.java:742]: Updated. PTAL @Jason918

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java (hunk 4 above)
Comment (Jason918) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java:993]: Please verify this if you have time. If there is a problem, we can fix them in another PR

**Review by Jason918 — APPROVED** [review]

**Comment by Jason918** [comment]
Can you help open a new PR to branch-2.10? There are a lot conflict when cherry-pick directly. @AnonHxy

**Comment by eolivelli** [comment]
@Jason918 we cannot add new APIs on released branches

**Comment by Jason918** [comment]
> @Jason918 we cannot add new APIs on released branches

Sure, I will remove the label.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 6]
```diff
@@ -601,6 +602,60 @@ private CompletableFuture<Map<String, String>> getPropertiesAsync() {
...
+                if (!opt.isPresent()) {
+                    throw new RestException(Status.NOT_FOUND,
+                        getTopicNotFoundErrorMessage(topicName.toString()));
+                }
+                ManagedLedger managedLedger = ((PersistentTopic) opt.get()).getManagedLedger();
+                managedLedger.asyncSetProperties(properties, new AsyncCallbacks.UpdatePropertiesCallback() {
+
+                    @Override
+                    public void updatePropertiesComplete(Map<String, String> properties, Object ctx) {
+                        managedLedger.getConfig().getProperties().putAll(properties);
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:646]: `managedLedger.getConfig().getProperties()` can return NPE if the topic is from the older code without the `properties`.

Comment (AnonHxy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:646]: Thanks for you to fix it!

**Pull request merged**